packages/apollo-datasource-rest/src/connectors/memcached.ts

```diff
@@ -0,0 +1,79 @@
+import { KeyValueCache } from '../KeyValueCache';
+import Memcached from 'memcached';
+import { promisify } from 'util';
+
+export default class MemcachedKeyValueCache implements KeyValueCache {
+  readonly client;
+  readonly defaultSetOptions = {
+    ttl: 300,
+    tags: [],
+  };
+  // might have to deal with issues of overflow, use BigInt?
+  private logicalClock = 0;
+
+  constructor(serverLocation: Memcached.Location, options?: Memcached.options) {
+    this.client = new Memcached(serverLocation, options);
+    // promisify client calls for convenience
+    this.client.get = promisify(this.client.get).bind(this.client);
+    this.client.getMulti = promisify(this.client.getMulti).bind(this.client);
+    this.client.set = promisify(this.client.set).bind(this.client);
+    this.client.flush = promisify(this.client.flush).bind(this.client);
+  }
+
+  async set(
+    key: string,
+    data: string,
+    options?: { ttl?: number; tags?: string[] },
+  ): Promise<void> {
+    const { ttl, tags } = Object.assign({}, this.defaultSetOptions, options);
+
+    // augment data with tags and version
+    const version = ++this.logicalClock;
+    const payload = {
+      v: version,
+      d: data,
+      t: tags,
+    };
+
+    await this.client.set(key, JSON.stringify(payload), ttl);
```

Comment:
Instead of serializing tag metadata together with the key, we could also store metadata in separate namespaced keys:
`set('key1', 'value1', ['tag1', 'tag2'])` gets stored like this: 
```
key1: value1
meta-tag-key1: 'tag1'|'tag2' 
meta-key-tag1: 'key1'
meta-key-tag2: 'key1'
```